agent/consul/state/catalog.go

```diff
@@ -3491,13 +3502,42 @@ func terminatingConfigGatewayServices(
 			CertFile:    svc.CertFile,
 			CAFile:      svc.CAFile,
 			SNI:         svc.SNI,
+			Kind:        kind,
 		}
 
 		gatewayServices = append(gatewayServices, mapping)
 	}
 	return false, gatewayServices, nil
 }
 
+func GatewayServiceKind(tx ReadTxn, name string, entMeta *acl.EnterpriseMeta) (structs.GatewayServiceKind, error) {
+	serviceIter, err := tx.First(tableServices, indexService, Query{
+		Value:          name,
+		EnterpriseMeta: *entMeta,
+	})
+	if err != nil {
+		return structs.GatewayServiceKindUnknown, err
+	}
+	if serviceIter != nil {
+		return structs.GatewayServiceKindService, err
+	}
+
+	_, entry, err := configEntryTxn(tx, nil, structs.ServiceDefaults, name, entMeta)
+	if err != nil {
+		return structs.GatewayServiceKindUnknown, err
+	}
+	if entry != nil {
+		sd, ok := entry.(*structs.ServiceConfigEntry)
+		if !ok {
+			return structs.GatewayServiceKindUnknown, fmt.Errorf("invalid config entry type %T", entry)
+		}
+		if sd.Destination != nil {
+			return structs.GatewayServiceKindDestination, nil
+		}
```

Comment:
```suggestion
		}
		return structs.GatewayServiceKindService, nil
```
Would this be appropriate? Or are you elsewhere going to interpret "unknown" as "service"?

Reply:
It will be dealt with in the watch receiving part https://github.com/hashicorp/consul/pull/13196
Once this PR got merged we will update https://github.com/hashicorp/consul/pull/13196 to have the right logic.

Reply:
The reason I added the unknown kind is to differentiate between existing services/destinations and references in the tgtw to non existing, it will help for debugging purpose.